File: apps/desktop/src/renderer/thinking-display-preferences.ts
/**
 * Thinking on the Feed has exactly two presentation effects:
 *
 * 1. Tip (`reasoning-stage`) — Summary lifecycle: shimmer label, tip only,
 *    removed when tools/正文 supersede. Used by OpenAI summary always, and by
 *    raw thinking when mode is `ephemeral` (阅后即焚).
 * 2. Card (`thinking`) — Collapsible ThinkingBlock. Used when mode is
 *    `collapsed` / `expanded` (default open state differs).
 *
 * Preference is applied in the projection display pipeline, not by hiding
 * ThinkingBlock after the fact.
 */
export type ThinkingDisplayMode = "ephemeral" | "collapsed" | "expanded";

export interface ThinkingDisplayPreferences {
  mode: ThinkingDisplayMode;
}

export const THINKING_DISPLAY_MODES: readonly ThinkingDisplayMode[] = [
  "ephemeral",
  "collapsed",
  "expanded",
] as const;

export const DEFAULT_THINKING_DISPLAY_PREFERENCES: ThinkingDisplayPreferences = {
  mode: "ephemeral",
};

export const THINKING_DISPLAY_STORAGE_KEY = "eco.thinking-display-preferences";
export const THINKING_DISPLAY_CHANGE_EVENT = "eco:thinking-display-change";

export function isThinkingDisplayMode(value: unknown): value is ThinkingDisplayMode {
  return value === "ephemeral" || value === "collapsed" || value === "expanded";
}

/** Tip path (Summary machinery), not a retained thinking card. */
export function thinkingModeUsesEphemeralTip(mode: ThinkingDisplayMode): boolean {
  return mode === "ephemeral";
}

/** Default expand state when the card path is used. */
export function thinkingModeDefaultExpanded(mode: ThinkingDisplayMode): boolean {
  return mode === "expanded";
}

export function normalizeThinkingDisplayPreferences(value: unknown): ThinkingDisplayPreferences {
  const candidate = value && typeof value === "object" ? (value as Record<string, unknown>) : {};
  if (isThinkingDisplayMode(candidate.mode)) {
    return { mode: candidate.mode };
  }
  // Migrate pre-mode boolean: preserve explicit collapsed/expanded; do not map to ephemeral.
  if (typeof candidate.thinkingContentDefaultExpanded === "boolean") {
    return {
      mode: candidate.thinkingContentDefaultExpanded ? "expanded" : "collapsed",
    };
  }
  return { ...DEFAULT_THINKING_DISPLAY_PREFERENCES };
}

export function readStoredThinkingDisplayPreferences(): ThinkingDisplayPreferences {
  try {
    const stored = localStorage.getItem(THINKING_DISPLAY_STORAGE_KEY);
    return stored
      ? normalizeThinkingDisplayPreferences(JSON.parse(stored))
      : { ...DEFAULT_THINKING_DISPLAY_PREFERENCES };
  } catch {
    return { ...DEFAULT_THINKING_DISPLAY_PREFERENCES };
  }
}

export function persistThinkingDisplayPreferences(preferences: ThinkingDisplayPreferences): void {
  const normalized = normalizeThinkingDisplayPreferences(preferences);
  try {
    localStorage.setItem(THINKING_DISPLAY_STORAGE_KEY, JSON.stringify(normalized));
  } catch {
    // Preference still applies in-memory when storage is unavailable.
  }
  window.dispatchEvent(
    new CustomEvent<ThinkingDisplayPreferences>(THINKING_DISPLAY_CHANGE_EVENT, {
      detail: normalized,
    }),
  );
}

File: apps/mobile/lib/core/preferences/thinking_display_preferences.dart
/// Thinking on the Feed has exactly two presentation effects (aligned with
/// desktop `thinking-display-preferences.ts`):
///
/// 1. Tip (`reasoningStage`) — Summary lifecycle: shimmer label, tip only,
///    removed when tools/正文 supersede. Used by OpenAI summary always, and by
///    raw thinking when mode is [ThinkingDisplayMode.ephemeral] (阅后即焚).
/// 2. Card (`thinking`) — Collapsible thinking tile. Used when mode is
///    [ThinkingDisplayMode.collapsed] / [ThinkingDisplayMode.expanded].
enum ThinkingDisplayMode {
  ephemeral,
  collapsed,
  expanded;

  static const storageKey = 'eco.thinking-display-preferences';

  static const all = ThinkingDisplayMode.values;

  static ThinkingDisplayMode? tryParse(String? raw) {
    return switch (raw) {
      'ephemeral' => ThinkingDisplayMode.ephemeral,
      'collapsed' => ThinkingDisplayMode.collapsed,
      'expanded' => ThinkingDisplayMode.expanded,
      _ => null,
    };
  }

  String get storageValue => name;

  /// Tip path (Summary machinery), not a retained thinking card.
  bool get usesEphemeralTip => this == ThinkingDisplayMode.ephemeral;

  /// Default expand state when the card path is used.
  bool get defaultExpanded => this == ThinkingDisplayMode.expanded;
}

const defaultThinkingDisplayMode = ThinkingDisplayMode.ephemeral;

ThinkingDisplayMode normalizeThinkingDisplayMode(Object? value) {
  if (value is ThinkingDisplayMode) return value;
  if (value is String) {
    return ThinkingDisplayMode.tryParse(value) ?? defaultThinkingDisplayMode;
  }
  if (value is Map) {
    final mode = ThinkingDisplayMode.tryParse(value['mode']?.toString());
    if (mode != null) return mode;
    // Migrate pre-mode boolean: preserve explicit collapsed/expanded.
    final legacy = value['thinkingContentDefaultExpanded'];
    if (legacy is bool) {
      return legacy
          ? ThinkingDisplayMode.expanded
          : ThinkingDisplayMode.collapsed;
    }
  }
  return defaultThinkingDisplayMode;
}
